pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -0,0 +1,154 @@
+package callbacks
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	util_xds "github.com/kumahq/kuma/pkg/util/xds"
+)
+
+// DataplaneCallbacks are XDS callbacks that keep the context of Kuma Dataplane.
+// In the ideal world we could assume that one Dataplane has one xDS stream.
+// Due to race network latencies etc. there might be a situation when one Dataplane has many xDS streams for the short period of time.
+// Those callbacks helps us to deal with such situation.
+//
+// Keep in mind that it does not solve many xDS streams across many instances of the Control Plane.
+// If there are many instances of the Control Plane and Dataplane reconnects, there might be an old stream
+// in one instance of CP and a new stream in a new instance of CP.
+//
+// Those callbacks may be also used with SDS. In case of SDS, at this moment Envoy creates many SDS streams to the Control Plane.
+type DataplaneCallbacks interface {
+	// OnStreamConnected is executed when Dataplane is connected to the control plane (go-control-plane's OnStreamOpen)
+	// and Dataplane executes the first request (go-control-plane's OnStreamRequest, only in this phase we can extract Dataplane and DataplaneMetadata).
+	// If there are many xDS stream this callback is executed for every new xDS stream.
+	OnStreamConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+	// OnFirstStreamConnected is similar to OnStreamConnected but if there are many xDS streams, it is only executed for the first stream.
+	OnFirstStreamConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
```

Comment:
When I read the interface for the first time I was thinking about the order of calls:
1. OnFirstStreamConnected
2. OnStreamConnected
3. OnStreamDisconnected
4. OnLastStreamDisconnected

In fact I think you have 1 and 2 switched. I'm not convinced here, maybe it's fine.

Also, when you see 2 methods:
- OnFirstStreamConnected
- OnStreamConnected

You may have an idea that method `OnStreamConnected` won't be called in case `OnFirstStreamConnected` was called.

And I don't really like the fact that we don't support many streams per Proxy, but at the same time this interface tells you the opposite.

Maybe we should rename the methods like:
- OnFirstStreamConnected -> OnProxyConnected
- OnStreamConnected -> OnProxyReconnected
- OnStreamConnected -> just drop it
- OnLastStreamDisconnected -> OnProxyDisconnected

WDYT?

Reply:
Thank you!! I was struggling a bit with naming here. I very much like `OnProxyConnected`, `OnProxyReconnected`, `OnProxyDisconnected`